SI-7202 tighten name mangling.

vigdorchik opened the pull request:
Following the discussion at https://groups.google.com/forum/?fromgroups=#!searchin/scala-internals/mangling/scala-internals/GHC3HJTegmw/5py5GH7Rrm8J
I'm re-submitting the change against master.

When the compiler wants to make private symbol visible outside its
declaring class, it also changes the name to make it unique to avoid
accidental overriding.

The current scheme prepends fullName of the owner class, which could
still result in overriding if the derived class is nested in a value
definition.

To avoid such scenario, name mangling prepends not only the names of
the enclosing classes, but also '$' characters for val owners that
aren't synthetic.

This scheme wouldn't be enough if the base class were local as well,
and the derived class saw it with a non-trivial prefix type; however
scala doesn't allow it.

Making this change required unification between normal 'fullName'
and the one used for name mangling. Using the opportunity, I also
clean the implementation of fullName to make recurring local, that
is a speedup over the late-bound call currently.

paulp commented on src/reflect/scala/reflect/internal/StdNames.scala:
```diff
@@ -341,12 +342,20 @@ trait StdNames {
+      @tailrec
+      def enclosing(sym: Symbol): Symbol = {
+        val owner = sym.effectiveOwner
+        if (owner.isClass || owner.isVal && !owner.name.containsChar(' ')) owner else enclosing(owner)
```
(owner.name containsChar ' ') is a huge abstraction failure. It's going to be hard enough to extinguish the "trailing whitespace" name without having to find places like this which depend on that implementation. If you had to check agains the name, you should be calling "isLocalName" or something like that; but I can't see why you'd even go to the name. You have the symbol, test the symbol for something more meaningful than its name.

paulp commented on src/reflect/scala/reflect/internal/Symbols.scala:
```diff
@@ -1033,16 +1033,21 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
-      else if (owner.isEffectiveRoot) name
-      else ((effectiveOwner.enclClass.fullNameAsName(separator) append separator): Name) append name
-    )
+    def fullNameAsName(separator: Char, enclosure: Symbol => Symbol = (_.effectiveOwner.enclClass)): Name = {
```
Both the default argument and the closure argument are unnecessary and serve to obscure the actual ways the method is used. Two variations do not a closure justify.

vigdorchik replied:
I thught we want to unify both methods, the usual fullName variant and the
one used for name mangling?

On Fri, Mar 29, 2013 at 2:43 PM, Paul Phillips notifications@github.comwrote:

> In src/reflect/scala/reflect/internal/Symbols.scala:
> 
> > @@ -1033,16 +1033,21 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
> >       */
> >      final def fullName(separator: Char): String = fullNameAsName(separator).toString
> > -    /*\* Doesn't drop package objects, for those situations (e.g. classloading)
> > -     \*  where the true path is needed.
> > -     */
> > -    private def fullNameInternal(separator: Char): Name = (
> > -      if (isRoot || isRootPackage || this == NoSymbol) name
> > -      else if (owner.isEffectiveRoot) name
> > -      else ((effectiveOwner.enclClass.fullNameAsName(separator) append separator): Name) append name
> > -    )
> > -    def fullNameAsName(separator: Char, enclosure: Symbol => Symbol = (_.effectiveOwner.enclClass)): Name = {
> 
> Both the default argument and the closure argument are unnecessary and
> serve to obscure the actual ways the method is used. Two variations do not
> a closure justify.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/scala/scala/pull/2334/files#r3584553
> .

paulp commented on src/reflect/scala/reflect/internal/Symbols.scala:
```diff
@@ -1033,16 +1033,21 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
-      else ((effectiveOwner.enclClass.fullNameAsName(separator) append separator): Name) append name
-    )
+    def fullNameAsName(separator: Char, enclosure: Symbol => Symbol = (_.effectiveOwner.enclClass)): Name = {
+      def recur(sym: Symbol): Name =
```
In the interests of not proliferating standards, I've been calling methods like this "loop" for years. I would rather see them all called "loop" than see "loop", "recur", ...

```
% ack --no-filename --no-group '\bdef loop\b' src/ |wc -l
      74
```

A couple of those are comments, but it's still a lot.

paulp commented on src/reflect/scala/reflect/internal/StdNames.scala:
```diff
@@ -273,6 +273,7 @@ trait StdNames {
     val DO_WHILE_PREFIX               = "doWhile$"
     val EVIDENCE_PARAM_PREFIX         = "evidence$"
     val EXCEPTION_RESULT_PREFIX       = "exceptionResult"
+    val EXPAND_FULL_NAME_SEPARATOR    = '$'
```
It is incredibly unlikely to me that this won't cause problems and make name mangling harder to fix. What if there are consecutive vals in the owner chain?

vigdorchik replied:
On Fri, Mar 29, 2013 at 3:01 PM, Paul Phillips notifications@github.comwrote:

> In src/reflect/scala/reflect/internal/StdNames.scala:
> 
> > @@ -273,6 +273,7 @@ trait StdNames {
> >      val DO_WHILE_PREFIX               = "doWhile$"
> >      val EVIDENCE_PARAM_PREFIX         = "evidence$"
> >      val EXCEPTION_RESULT_PREFIX       = "exceptionResult"
> > -    val EXPAND_FULL_NAME_SEPARATOR    = '$'
> 
> It is incredibly unlikely to me that this won't cause problems and make
> name mangling harder to fix. What if there are consecutive vals in the
> owner chain?
> 
> Vals are inserted with their names, so I don't see a problm here.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/scala/scala/pull/2334/files#r3584646
> .

paulp replied:
What if the val is called <*> ?

paulp replied:
If it's not clear what I mean yet, how do you avoid a $ appearing next to another $ and thereby being decoded as EXPAND_SEPARATOR_STRING instead of EXPAND_FULL_NAME_SEPARATOR ?

vigdorchik replied:
On Fri, Mar 29, 2013 at 3:58 PM, Paul Phillips notifications@github.comwrote:

> In src/reflect/scala/reflect/internal/StdNames.scala:
> 
> > @@ -273,6 +273,7 @@ trait StdNames {
> >      val DO_WHILE_PREFIX               = "doWhile$"
> >      val EVIDENCE_PARAM_PREFIX         = "evidence$"
> >      val EXCEPTION_RESULT_PREFIX       = "exceptionResult"
> > -    val EXPAND_FULL_NAME_SEPARATOR    = '$'
> 
> What if the val is called <*> ?
> 
> Isn't it as the class called '*'? So it should work, as the names are
> encoded.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/scala/scala/pull/2334/files#r3584995
> .

vigdorchik replied:
On Fri, Mar 29, 2013 at 4:00 PM, Paul Phillips notifications@github.comwrote:

> In src/reflect/scala/reflect/internal/StdNames.scala:
> 
> > @@ -273,6 +273,7 @@ trait StdNames {
> >      val DO_WHILE_PREFIX               = "doWhile$"
> >      val EVIDENCE_PARAM_PREFIX         = "evidence$"
> >      val EXCEPTION_RESULT_PREFIX       = "exceptionResult"
> > -    val EXPAND_FULL_NAME_SEPARATOR    = '$'
> 
> If it's not clear what I mean yet, how do you avoid a $ appearing next to
> another $ and thereby being decoded as EXPAND_SEPARATOR_STRING instead of
> EXPAND_FULL_NAME_SEPARATOR ?
> 
> That is ensured by the non-emptiness of the name, with  local suffix
> dropped.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/scala/scala/pull/2334/files#r3585007
> .

paulp replied:
What does "the names are encoded" mean to you?

paulp commented:
So do I understand that the elements being purged from the owner chain (replaced with a $) are the local dummies? That's an improvement over what I was picturing, but it wasn't clear from the code, comments, or commit message. Can you add some emphasis that "local dummies" are the key participants in this?

paulp commented on src/reflect/scala/reflect/internal/Symbols.scala:
```diff
@@ -1033,16 +1033,24 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
-      else if (owner.isEffectiveRoot) name
-      else ((effectiveOwner.enclClass.fullNameAsName(separator) append separator): Name) append name
-    )
+    private def properEnclClass = effectiveOwner.enclClass
```
Does this amount to "effectiveOwner, but skip local dummies" the same way that effectiveOwner is "owner, but skip package objects" ? Like dealiasWidenAndChaseUpperBounds, further steps down this road threaten a compositional nightmare. Is there anyone who can better justify the need for and operation of local dummies in the first place? It would be far better if we didn't have this fluctuating owner chain which requires manual management of what should be jumped.

vigdorchik replied:
On Fri, Mar 29, 2013 at 4:38 PM, Paul Phillips notifications@github.comwrote:

> In src/reflect/scala/reflect/internal/Symbols.scala:
> 
> > @@ -1033,16 +1033,24 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
> >       */
> >      final def fullName(separator: Char): String = fullNameAsName(separator).toString
> > -    /*\* Doesn't drop package objects, for those situations (e.g. classloading)
> > -     \*  where the true path is needed.
> > -     */
> > -    private def fullNameInternal(separator: Char): Name = (
> > -      if (isRoot || isRootPackage || this == NoSymbol) name
> > -      else if (owner.isEffectiveRoot) name
> > -      else ((effectiveOwner.enclClass.fullNameAsName(separator) append separator): Name) append name
> > -    )
> > -    private def properEnclClass = effectiveOwner.enclClass
> 
> Does this amount to "effectiveOwner, but skip local dummies" the same way
> that effectiveOwner is "owner, but skip package objects" ? Like
> dealiasWidenAndChaseUpperBounds, further steps down this road threaten a
> compositional nightmare. Is there anyone who can better justify the need
> for and operation of local dummies in the first place? It would be far
> better if we didn't have this fluctuating owner chain which requires manual
> management of what should be jumped.
> 
> I think this coment is obsolete, isn't it?
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/scala/scala/pull/2334/files#r3585218
> .

paulp commented:
Sorry, I'm sure it seems like I'm piling on this PR, but I have very good reason for being scared of changes to name mangling.

vigdorchik commented:
I extracted vals for enclosure parameters in two places for performance.
As for dummy vs. local owners, the name of both of which contain spaces, it's only the former that has to be excluded; since those represent code blocks that are embedded in class primary constructors, they can't be referenced.
As for local owners, they need to be included, and their whitespace suffix is dropped at the call site.
I added respective comment to the code.

paulp commented:
On the closure, performance is only part of the problem. You are bringing the full weight of absolute generality to bear on a problem space which fits into a boolean. It is swatting a fly with a sledgehammer. It is harder to understand, harder to modify, and easier to break. Perhaps there will come a day when the problem space no longer fits into a boolean (though it will surely fit into a byte for a long time yet) but that day is not here.

vigdorchik committed:
SI-7202 tighten name mangling.

When the compiler wants to make private symbol visible outside its
declaring class, it also changes the name to make it unique to avoid
accidental overriding.

The current scheme prepends fullName of the owner class, which could
still result in overriding if the derived class is nested in a value
definition.

To avoid such scenario, name mangling prepends not only the names of
the enclosing classes, but also '!' characters for non-class owners.
'!' is chosen as a valid character in binary name, that is escaped
in scala source code.

Adding a single character is enough since any local classes defined
in mangled term owners are not visible outside, so any inadvertent
overriding between two locals is not possible.

vigdorchik commented:
@paulp Please take another look. I decided the similarity to full name
doesn't deserve closure indeed, nor is it pretty to have a flag.
Instead I introduce a second local function. Also I included a comment with
a proof the new scheme is sound.

On Sat, Mar 30, 2013 at 8:28 PM, Paul Phillips notifications@github.comwrote:

> On the closure, performance is only part of the problem. You are bringing
> the full weight of absolute generality to bear on a problem space which
> fits into a boolean. It is swatting a fly with a sledgehammer. It is harder
> to understand, harder to modify, and easier to break. Perhaps there will
> come a day when the problem space no longer fits into a boolean (though it
> will surely fit into a byte for a long time yet) but that day is not here.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/scala/scala/pull/2334#issuecomment-15677226
> .

paulp commented on src/reflect/scala/reflect/internal/StdNames.scala:
```diff
@@ -341,8 +341,27 @@ trait StdNames {
+        else
+          // The correctness argument to insert a single '$' for non-class owners goes as follows:
+          // 1) It's impossible to inadvertently override a symbol whose owner chain doesn't contain
+          // non-class symbols, since this has at most one consecutive '$' in its prefix.
```
I can't understand 1) at all. Can you try another way of expressing the part before the comma? And how is it that the part after the comma ("since ...") is given as an axiom? How do we know this has at most one consecutive '$' in its prefix?

paulp commented on src/reflect/scala/reflect/internal/StdNames.scala:
```diff
@@ -341,8 +341,27 @@ trait StdNames {
+          // 2) For two symbols with the same expanded name that both have non-class symbols in their owner chain,
+          // let us consider the declaring classes and let's find the longest prefix of expanded
+          // name such that both classes have same owners corresponding to the prefix. Then the suffix has to
+          // start with "$$$" and corresponding non-class owners have to be different. Since scala only allows
```
What is the logic which leads to "then the suffix has to start with $$$" ? I think both 1) and 2) need a detailed example, probably more than one.

JamesIry commented:
@vigdorchik, @paulp what's the current status on this PR?

vigdorchik commented:
On Wed, Apr 10, 2013 at 10:36 PM, James Iry notifications@github.comwrote:

> @vigdorchik https://github.com/vigdorchik, @paulphttps://github.com/paulpwhat's the current status on this PR?
> 
> I'm curious too...
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/scala/scala/pull/2334#issuecomment-16193253
> .

paulp commented:
I am very, very reluctant to proceed with this. I know to moral certainty that adding new complexity to name mangling is going to pay me back 100x over in the process of dismantling name mangling so it can be done properly. Is there a specific driver for the attempt to fix this? By which I mean, somebody's client or it is blocking their own work with no workaround? If push comes to shove I will have to drop the other things I should be doing to muster the campaign against fiddling with name mangling over corner cases.

Basically, DNLGTM and is not likely to without a LOT more tests. I don't just want to see tests for what you're trying to fix; I want to see tests for everything which isn't supposed to be quietly changing behavior or losing some quality which is still in its very tenuous grasp, and I don't want to see my small set of tools for fixing name mangling further curtailed by seeing new characters introduced as delimiters. You will have to take ! from my dead hands. If we introduce ! it should be as part of a completely thought out and comprehensive fix. Not to fix one lonely corner case.
